tmpl/formats/epub.go

```diff
@@ -3,31 +3,129 @@ package formats
 import (
 	"archive/zip"
 	"context"
+	"fmt"
+	"github.com/rushsteve1/mangadex-opds/shared"
+	"github.com/rushsteve1/mangadex-opds/tmpl"
+	"golang.org/x/sync/errgroup"
 	"io"
+	"log/slog"
+	"os"
+	"path"
 
 	"github.com/rushsteve1/mangadex-opds/models"
 )
 
 // WriteEpub will write an EPUB file for the current [Chapter] to the given [io.Writer].
 func WriteEpub(ctx context.Context, c *models.Chapter, w io.Writer) (err error) {
-	// A CBZ is almost a valid EPUB so we can use that as the base
-	WriteCBZ(ctx, c, w)
+	z := zip.NewWriter(w)
 
-	// err = c.writeEpubMetadata(z)
+	err = z.SetComment(c.FullTitle())
+	if err != nil {
+		return err
+	}
 
-	return err
-}
+	err = z.AddFS(os.DirFS("tmpl/formats/epub"))
+	if err != nil {
+		return err
+	}
+
+	imgUrls, err := c.FetchImageURLs(ctx)
+	if err != nil {
+		return err
+	}
 
-// TODO implement this function
-// - Epub metadata
-// - Epub structure
-func writeEpubMetadata(c *models.Chapter, z *zip.Writer) (err error) {
-	w, err := z.Create("META-INF/container.xml")
+	w, err = z.Create("content.opf")
 	if err != nil {
 		return err
 	}
 
-	_, err = w.Write([]byte("TODO"))
+	err = tmpl.ContentOPF(c, w)
+	if err != nil {
+		return err
+	}
+
+	w, err = z.Create("toc.ncx")
+	if err != nil {
+		return err
+	}
+
+	err = tmpl.TocNCX(c, w)
+	if err != nil {
+		return err
+	}
+
+	imgChan := make(chan tmpl.ChapterImage)
+	doneChan := make(chan error)
+
+	// Fetch and add the image files in parallel
+	go func() {
```

Comment:
I feel like there might be a way to simplify this logic and combine it with the CBZ code, but that's a later TODO item.